portality/bll/services/article.py

```diff
@@ -0,0 +1,375 @@
+        :return: a report on the state of the import: {success: x, fail: x, update: x, new: x, shared: [], unowned: [], unmatched: []}
+        """
+        # first validate the incoming arguments to ensure that we've got the right thing
+        argvalidate("batch_create_article", [
+            {"arg": articles, "instance" : list, "allow_none" : False, "arg_name" : "articles"},
+            {"arg": account, "instance" : models.Account, "allow_none" : False, "arg_name" : "account"},
+            {"arg" : duplicate_check, "instance" : bool, "allow_none" : False, "arg_name" : "duplicate_check"},
+            {"arg" : merge_duplicate, "instance" : bool, "allow_none" : False, "arg_name" : "merge_duplicate"},
+            {"arg" : limit_to_account, "instance" : bool, "allow_none" : False, "arg_name" : "limit_to_account"}
+        ], exceptions.ArgumentException)
+
+        # 1. dedupe the batch
+        if duplicate_check:
+            batch_duplicates = self._batch_contains_duplicates(articles)
+            if batch_duplicates:
+                report = {"success" : 0, "fail" : len(articles), "update" : 0, "new" : 0, "shared" : [], "unowned" : [], "unmatched" : []}
+                raise exceptions.IngestException(message=Messages.EXCEPTION_ARTICLE_BATCH_DUPLICATE, result=report)
+
+        # 2. check legitimate ownership
+        success = 0
+        fail = 0
+        update = 0
+        new = 0
+        all_shared = set()
+        all_unowned = set()
+        all_unmatched = set()
+
+        for article in articles:
+            result = self.create_article(article, account, duplicate_check=duplicate_check, merge_duplicate=merge_duplicate, limit_to_account=limit_to_account, dry_run=True)
+            success += result.get("success", 0)
+            fail += result.get("fail", 0)
+            update += result.get("update", 0)
+            new += result.get("new", 0)
+            all_shared.update(result.get("shared", set()))
+            all_unowned.update(result.get("unowned", set()))
+            all_unmatched.update(result.get("unmatched", set()))
+
+        report = {"success" : success, "fail" : fail, "update" : update, "new" : new, "shared" : all_shared, "unowned" : all_unowned, "unmatched" : all_unmatched}
+
+        # if there were no failures in the batch, then we can do the save
+        if fail == 0:
+            for i in xrange(len(articles)):
+                block = i == len(articles) - 1
+                # block on the final save, so that when this method returns, all articles are
+                # available in the index
+                articles[i].save(blocking=block)
+
+            # return some stats on the import
+            return report
+        else:
+            raise exceptions.IngestException(message=Messages.EXCEPTION_ARTICLE_BATCH_FAIL, result=report)
+
+
+    def _batch_contains_duplicates(self, articles):
```

Comment:
```suggestion
    def _batch_contains_duplicates(self, articles):
        dois = [a.get_normalised_doi() for a in articles]
        dup_doi = len(set(dois)) != len(dois)
        
        fts = [a.get_normalised_doi() for a in articles]
        dup_ft = len(set(fts)) != len(fts)
        
        return dup_doi or dup_ft
```

hmm, mine always loops through articles twice, yours could halt early too. You win!